templates/android/library/src/main/java/io/appwrite/services/Realtime.kt.twig

```diff
@@ -0,0 +1,165 @@
+import kotlinx.coroutines.Dispatchers.IO
+import okhttp3.*
+import okhttp3.internal.concurrent.TaskRunner
+import okhttp3.internal.ws.RealWebSocket
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+class Realtime(client: Client) : Service(client), CoroutineScope {
+
+    private val job = Job()
+
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.Main + job
+
+    private companion object {
+        private const val DEBOUNCE_MILLIS = 1L
+
+        private var socket: RealWebSocket? = null
+        private var channelCallbacks = mutableMapOf<String, MutableCollection<(Any) -> Unit>>()
+        private var errorCallbacks = mutableSetOf<(RealtimeError) -> Unit>()
+
+        private var subCallDepth = 0
+    }
+
+    private fun createSocket() {
+        val queryParamBuilder = StringBuilder()
+            .append("project=${client.config["project"]}")
+
+        channelCallbacks.keys.forEach {
+            queryParamBuilder
+                .append("&channels[]=$it")
+        }
+
+        val request = Request.Builder()
+            .url("${client.endPointRealtime}/realtime?$queryParamBuilder")
+            .build()
+
+        if (socket != null) {
+            closeSocket()
+        }
+
+        socket = RealWebSocket(
+            taskRunner = TaskRunner.INSTANCE,
+            originalRequest = request,
+            listener = {{ spec.title | caseUcfirst }}WebSocketListener(),
+            random = Random(),
+            pingIntervalMillis = client.http.pingIntervalMillis.toLong(),
+            extensions = null,
+            minimumDeflateSize = client.http.minWebSocketMessageToCompress
+        )
+
+        socket?.connect(client.http)
+    }
+
+    private fun closeSocket() {
+        socket?.close(RealtimeCodes.POLICY_VIOLATION.value, null)
+    }
+
+    fun subscribe(
+        vararg channels: String,
+        callback: (Any) -> Unit
+    ) : RealtimeSubscription {
+        channels.forEach {
+            if (!channelCallbacks.containsKey(it)) {
+                channelCallbacks[it] = mutableListOf(callback)
+                return@forEach
+            }
+            channelCallbacks[it]?.add(callback)
+        }
+
+        launch {
+            subCallDepth++
+            delay(DEBOUNCE_MILLIS)
+            if (subCallDepth == 1) {
+                createSocket()
+            }
+            subCallDepth--
+        }
+
+        return RealtimeSubscription { unsubscribe() }
+    }
+
+    fun unsubscribe() {
+        channelCallbacks = mutableMapOf()
+        errorCallbacks = mutableSetOf()
+        closeSocket()
+    }
+
+    fun doOnError(callback: (RealtimeError) -> Unit) {
+        errorCallbacks.add(callback)
+    }
+
+    private inner class {{ spec.title | caseUcfirst }}WebSocketListener : WebSocketListener() {
+
+        override fun onOpen(webSocket: WebSocket, response: Response) {
+            super.onOpen(webSocket, response)
+            print("WebSocket connected.")
```

Comment:
Do we continue to leave these log messages here ? If yes, I think it's better to replace print with `Log.i()`